src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp

```diff
@@ -40,91 +40,73 @@ using namespace std;
   }
 
-  BinnedSpectrum::BinnedSpectrum(float size, UInt spread, PeakSpectrum ps) :
-    bin_spread_(spread), bin_size_(size), bins_(), raw_spec_(ps)
+  BinnedSpectrum::~BinnedSpectrum()
   {
-    setBinning();
   }
 
-  BinnedSpectrum::BinnedSpectrum(const BinnedSpectrum& source) :
-    bin_spread_(source.getBinSpread()), bin_size_(source.getBinSize()), bins_(source.getBins()), raw_spec_(source.raw_spec_)
+  const SparseVector<float>& BinnedSpectrum::getBins() const
   {
+    return bins_;
   }
 
-  BinnedSpectrum::~BinnedSpectrum()
+  SparseVector<float>& BinnedSpectrum::getBins()
   {
+    return bins_;
   }
 
-  //accessors and operators see .h file
+  const std::vector<Precursor>& BinnedSpectrum::getPrecursors() const
+  {
+    return precursors_;
+  }
 
+  std::vector<Precursor>& BinnedSpectrum::getPrecursors()
+  {
+    return precursors_;
+  }
 
-  void BinnedSpectrum::setBinning()
+  void BinnedSpectrum::binSpectrum_(const PeakSpectrum& ps)
   {
-    if (raw_spec_.empty())
-    {
-      throw BinnedSpectrum::NoSpectrumIntegrated(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
-    }
-    bins_.clear();
+    // TODO: add assertion to check if spectrum ps is sorted by position or retrieve maximum differently
 
-    //make all necessary bins accessible
-    raw_spec_.sortByPosition();
-    bins_ = SparseVector<float>((UInt)ceil(raw_spec_.back().getMZ() / bin_size_) + bin_spread_, 0, 0);
+    const Size highest_index = floor(ps.back().getMZ() / bin_size_) + bin_spread_;
+    bins_ = SparseVector<float>((UInt)highest_index + 1, 0, 0);
 
-    //put all peaks into bins
-    UInt bin_number;
-    for (Size i = 0; i < raw_spec_.size(); ++i)
+    // put all peaks into bins
+    for (auto const & p : ps)
     {
-      //bin_number counted form 0 -> floor
-      bin_number = (UInt)floor(raw_spec_[i].getMZ() / bin_size_);
-      //e.g. bin_size_ = 1.5: first bin covers range [0,1.5] so peak at 1.5 falls in first bin (index 0)
-
-      if (raw_spec_[i].getMZ() / bin_size_ == (double)bin_number)
-      {
-        --bin_number;
-      }
+      // e.g.: bin_size_ = 1.5: first bin covers range [0, 1.5) so peak at 1.5 falls in second bin (index 1)
+      UInt idx = (UInt)floor(p.getMZ() / bin_size_);
 
-      //add peak to corresponding bin
-      bins_[bin_number] = bins_.at(bin_number) + raw_spec_[i].getIntensity();
+      // add peak to corresponding bin
+      bins_[idx] = bins_.at(idx) + p.getIntensity();
 
-      //add peak to neighboring binspread many
+      // add peak to neighboring bins
       for (Size j = 0; j < bin_spread_; ++j)
       {
-        bins_[bin_number + j + 1] = bins_.at(bin_number + j + 1) + raw_spec_[i].getIntensity();
-        // we are not in one of the first bins (0 to bin_spread)
-        //not working:  if (bin_number-j-1 >= 0)
-        if (bin_number >= j + 1)
+        bins_[idx + j + 1] = bins_.at(idx + j + 1) + p.getIntensity();
+        // prevent spreading over left boundaries
+        if (idx >= j + 1)
```

Comment:
why not just write `idx-j-1 >= 0`? 

also, why dont we check to go over the right boundary? what if `idx+j+1` is larger than the vector?

Reply:
good point .... its copy & pasted code from the old implementation